### PR TITLE
[GPU]: Slice op supports dynamic shapes

### PR DESCRIPTION
**Pull request by pkowalc1**
### Details:
 - Slice op on GPU supports dynamic shapes

### Tickets:
 - *CVS-129985*
 - *CVS-100235*

### CODE REVIEW

File: src/plugins/intel_gpu/tests/unit/test_cases/slice_gpu_test.cpp [hunk 1]
```diff
@@ -163,17 +308,26 @@ TYPED_TEST(SliceTest, bfzyx) {
...
+
+    this->RunAllTestCasesForParams(params);
+    params.is_input_dynamic = true;
+    params.is_start_dynamic = true;
+    params.is_step_dynamic = true;
+    params.is_stop_dynamic = true;
+    params.is_axes_dynamic = true;
+
     this->RunAllTestCasesForParams(params);
 }
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/tests/unit/test_cases/slice_gpu_test.cpp:332]: Could you please add functional tests as well? They are located here `src/plugins/intel_gpu/tests/functional/single_layer_tests/dynamic/`
And currently `StridedSlice` operation replaces `Slice` op during `SliceToStridedSlice` transformation, so you need to disable this transformation to keep slice op in ov::Model

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl [hunk 2]
```diff
@@ -4,33 +4,89 @@
...
+
+    long slice_step[INPUT0_DIMS];
+    long slice_start[INPUT0_DIMS];
+    #pragma unroll
+    for(int i = 0; i < INPUT0_DIMS; ++i) {
+        slice_step[i] = 1;
+        slice_start[i] = 0;
+    }
+
+    #pragma unroll
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:46]: nit: usually we use GPU Plugin's macro `unroll_for` for unrolling hint

File: src/plugins/intel_gpu/src/graph/slice.cpp [hunk 3]
```diff
@@ -10,6 +10,30 @@
...
+
+SliceKernelRefNeededInputs SliceKernelRefNeededInputs::Create(const slice_node& node) {
+    SliceKernelRefNeededInputs inputs;
+
+    const auto& node_inputs = node.get_dependencies();
+
+    const bool axes_in_runtime =
+        ((node_inputs.size() == InputIndices::kInputsNum) && !node_inputs[InputIndices::kAxes].first->is_constant());
+    const bool start_in_runtime = node_inputs[InputIndices::kStart].first->is_dynamic();
+    const bool step_in_runtime = node_inputs[InputIndices::kStep].first->is_dynamic();
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/slice.cpp:22]: This `is_dynamic()` method checks if program_node has any dynamic input/outputs _in terms of shape_ not the data, so if your intention was to check the data itself (if it's constant data or not), you need to use `!is_constant()`

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl [hunk 4]
```diff
@@ -4,33 +4,89 @@
...
+    for(int i = 0; i < AXES_BUFFER_SIZE; ++i) {
+        const long axis = axes_buff[i];
+        slice_step[axis] = step_buff[i];
+        slice_start[axis] = start_buff[i];
+    }
+
+    const long output_dim0 = get_global_id(0);
+    const long output_dim1 = get_global_id(1);
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:54]: It's better to use uint for `get_global_id()` and indexes calculations, since `long` calculation may be quite time consuming, but uint data range is enough for most of the cases

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl [hunk 5]
```diff
@@ -4,33 +4,89 @@
...
+        slice_begin_dim2 + output_dim2 * slice_step[2],
+        slice_begin_dim3 + output_dim3 * slice_step[3],
+        slice_begin_dim4 + output_dim4 * slice_step[4]);
 #endif
+
     output[output_index] = ACTIVATION(input[input_index], ACTIVATION_PARAMS);
 }
+
+#undef LOAD_BUFFER;
+#undef BRING_INTO_RANGE;
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:92]: nit: please add new line at the end of the file

File: src/plugins/intel_gpu/src/graph/impls/ocl/slice.cpp [hunk 6]
```diff
@@ -40,104 +39,149 @@ std::vector<std::int32_t> extractIntegerData(const data_node& node, const stream
...
-            auto transformed_axe = axes[axis] < 0 ? data_shape.size() + axes[axis] : axes[axis];
-            auto start = start_elts[axis];
-            auto end = end_elts[axis];
-            auto dim_size = data_shape[transformed_axe];
-            selected_start[transformed_axe] = std::max(std::min(start < 0 ? dim_size + start : start, dim_size - 1), 0);
-            selected_end[transformed_axe] = std::max(std::min(end < 0 ? dim_size + end : end, dim_size - 1), 0);
-            selected_step[transformed_axe] = step_elts[axis];
+        auto params = get_default_params<kernel_selector::slice_params>(impl_param, impl_param.is_dynamic());
+
+        OPENVINO_ASSERT(PrepareInput(arg,
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/impls/ocl/slice.cpp:94]: Could you please add meaningful assert message here please?

File: src/plugins/intel_gpu/tests/unit/test_cases/slice_gpu_test.cpp (hunk 1 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/tests/unit/test_cases/slice_gpu_test.cpp:332]: 1. Ok, will add some functional tests as well.
2. Disabling SliceToStridedSlice transformation should be done in separate mr

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl (hunk 2 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:46]: OK, fixed.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl (hunk 5 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:92]: Fixed.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl (hunk 4 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:54]: I thought about that. However, it is problematic IMO. First of all, the slice op support int64_t values(I also agree that it is pointless, but that's the definition of that op). So in general it might happen that somebody will one day use values that will fit only into int64. Since this is a reference implementation, it is suppose to work with all possible parameters, I have decided to keep everything in long, since it will always fit. Previously, slice op was silently down casting in64_t to in32, which was bad IMO. If we don't want to support int64_t, then it should be an assert, that  "you have int64_t input parameters, but gpu supports only int32_t". But that is the question of what is that reference implementation. 

About performance, I agree, it may happen that performance might be less - however it all depends on the hardware. In this particular case the degradation shouldn't be much(however, I haven't check it and my claims are based mostly on NVIDIA architectures), since the kernel is memory_bound, so in general case, the bottleneck of this kernel will be memory access pattern , not the offset calculation itself(assuming that you have enough blocks/threads to switch to - it is called high enough occupancy in NVIDIA terms - it should be true since I am talking about general case here, not some edge cases where data is small etc).

So to sum up, taking all the tradeoffs, I have assumed that it is more important for this kernel to be conformant with this slice op def, than being 'optimized', since this kernel is not optimized anyway:) The goal was to add support for dynamic shape, not to optimize it, so I tried to do it in way, that the performance in cases that were already supported(namely, static input) will not be seriously degraded.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/slice/slice_kernel_ref.cpp [hunk 7]
```diff
@@ -7,23 +7,72 @@
...
+    const std::string BUFF_PTR_NAME = name + "_buffer_ptr";
+    const auto jit_name_decorator = [](std::string name, size_t i) {
+        return name + "_VAL" + std::to_string(i);
+    };
+
+    if (compile_time_param.empty()) {
+        // Dynamic param:
+        const std::string type_str = ovElementTypeToOCLStr(type);
+        jit.AddConstant(
+            MakeJitConstant(BUFF_CONST_NAME, "__global const " + type_str + "* restrict " + BUFF_PTR_NAME + ","));
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/slice/slice_kernel_ref.cpp:55]: Got your point, then I'd suggest to replace `ov::element::Type_t` in slice_params with `kernel_selector::Datatype` (there is a function to convert ov data type to kernel_selector's data type: `kernel_selector::data_type to_data_type(data_types dt)`) and replace custom `std::string ovElementTypeToOCLStr` with an existing `std::string toCLType(Datatype dType)` func - just for unification
BTW, as of now GPU Plugin converts the most part of integers in the model to i32 data type (please see int_convert_precision_map in transformations pipeline), so kernel_selector::data_type would be enough to cover all supported cases

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl (hunk 4 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/slice_ref.cl:53]: @pkowalc1 yeah, absolutely agree with your point
Just some additional thoughts:
If there will be the real case where uint32_t data range wouldn't be enough, most likely the plugin will fail much earlier, because of the same limitations in the other kernels :)
As this slice op is going to replace strided_slice (and it is supposed to have at least on par or even better performance), some additional perf can be helpful
But personally I'm okay with long data type

File: src/plugins/intel_gpu/src/graph/impls/ocl/slice.cpp [hunk 8]
```diff
@@ -40,104 +39,152 @@ std::vector<std::int32_t> extractIntegerData(const data_node& node, const stream
...
+                                     SliceKernelRefNeededInputs::kStart,
+                                     params.compile_time_start,
+                                     params.start_data_type,
+                                     params.inputs),
+                        "[GPU][SliceOp]: Slice op definition does not contain start input!");
+
+        // NOTE: Stop input is not used by the slice kernel, as this information
+        // is implicitely passed with output shape.
+
+        OPENVINO_ASSERT(PrepareInput(arg,
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/impls/ocl/slice.cpp:104]: As this `PrepareInput` function is doing something, not only checking, I think it would be better to move it outside of OPENVINO_ASSERT, just to improve readability a little bit:
```
bool status = PrepareInput(..);
OPENVINO_ASSERT(success, "msg");

status = PrepareInput(..);
OPENVINO_ASSERT(success, "msg");
...
```

Comment (pkowalc1) [reply src/plugins/intel_gpu/src/graph/impls/ocl/slice.cpp:104]: Fixed, but in different way: assert is removed completely, since by def step and start may not be set at all and then default values should be applied.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/slice/slice_kernel_ref.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/slice/slice_kernel_ref.cpp:55]: Fixed.

About int32 vs int64 - it is up to you guys, I can make it to work with int32, but as I said, slice op by def should accept int64 data types as params.

**Review by sshlyapn — APPROVED** [review]
